pkg/vm/vm_core.go

```diff
@@ -1,15 +1,173 @@
+		if err != nil {
+			return err
+		}
+		returnPC := memory.NewMaybeRelocatableRelocatable(new_rel)
+
+		if !operands.Op0.IsEqual(returnPC) {
+			return &VirtualMachineError{"CantWriteReturnPc"}
+		}
+
+		returnFP := vm.RunContext.Fp
+		dstRelocatable, _ := operands.Dst.GetRelocatable()
+		if !returnFP.IsEqual(&dstRelocatable) {
+			return &VirtualMachineError{"CantWriteReturnFp"}
+		}
+	}
+
+	return nil
+}
+
+func (deduced *DeducedOperands) set_dst(value uint8) {
+	deduced.Operands = deduced.Operands | value
+}
+
+func (deduced *DeducedOperands) set_op0(value uint8) {
+	deduced.Operands = deduced.Operands | value<<1
+}
+
+func (deduced *DeducedOperands) set_op1(value uint8) {
+	deduced.Operands = deduced.Operands | value<<2
+}
+
+//------------------------
+//  virtual machines funcs
+// ------------------------
+
+func VmNew(run_context RunContext, current_step uint, segments_manager memory.MemorySegmentManager) VirtualMachine {
+	return VirtualMachine{
+		RunContext:  run_context,
+		CurrentStep: current_step,
+		Segments:    segments_manager,
+	}
+}
+
+func (vm *VirtualMachine) ComputeRes(instruction Instruction, op0 memory.MaybeRelocatable, op1 memory.MaybeRelocatable) (memory.MaybeRelocatable, error) {
+	switch instruction.ResLogic {
+	case ResOp1:
+		return op1, nil
+
+	case ResAdd:
+		maybe_rel, err := op0.AddMaybeRelocatable(op1)
+		if err != nil {
+			return memory.MaybeRelocatable{}, errors.New("adding maybe relocatable")
+		}
+		return maybe_rel, nil
+
+	case ResMul:
+		num_op0, m_type := op0.GetInt()
+		num_op1, other_type := op1.GetInt()
+		if m_type && other_type {
+			result := memory.NewMaybeRelocatableInt(num_op0.Felt.Add(num_op1.Felt))
+			return *result, nil
+		} else {
+			return memory.MaybeRelocatable{}, errors.New("ComputeResRelocatableMul")
+		}
+
+	case ResUnconstrained:
+		return memory.MaybeRelocatable{}, nil
+	}
+	return memory.MaybeRelocatable{}, nil
+}
+
+func (vm *VirtualMachine) ComputeOperands(instruction Instruction) (Operands, OperandsAddresses, DeducedOperands, error) {
+
+	dst_addr, err := vm.RunContext.ComputeDstAddr(instruction)
+	if err != nil {
+		return Operands{}, OperandsAddresses{}, DeducedOperands{}, errors.New("FailtToComputeDstAddr")
+	}
+	dst_op, _ := vm.Segments.Memory.Get(dst_addr)
+
+	op0_addr, err := vm.RunContext.ComputeOp0Addr(instruction)
+	if err != nil {
+		return Operands{}, OperandsAddresses{}, DeducedOperands{}, errors.New("FailtToComputeOp0Addr")
+	}
+	op0_op, _ := vm.Segments.Memory.Get(op0_addr)
+
+	op1_addr, err := vm.RunContext.ComputeOp1Addr(instruction, *op0_op)
+	if err != nil {
+		return Operands{}, OperandsAddresses{}, DeducedOperands{}, errors.New("FailtToComputeOp1Addr")
+	}
+	op1_op, _ := vm.Segments.Memory.Get(op1_addr)
+
+	deduced_operands := DeducedOperands{Operands: 0}
+	res, err := vm.ComputeRes(instruction, *op0_op, *op1_op)
+
+	accesed_addresses := OperandsAddresses{
+		Dst_addr: dst_addr,
+		Op0_addr: op0_addr,
+		Op1_addr: op1_addr,
+	}
```

Comment:
We can ignore this at this stage of the vm

Reply:
must be deleted? as is going to be used and does not broke anything, maybe we can leave it

Reply:
Its dead code till we get to proof_mode/starknet_in_go(?)

Reply:
i delete it then